docs/historical/mh-jun-1982/man/mail.1

.TH MAIL 1 RAND
.SH NAME
mail \- send or read mail
.SH SYNOPSIS
.B mail
[[\fB\-subject\fR\ subject] [\fB\-cc\ \fRname ...] [\fB\-help\fR] user ...]
.SH DESCRIPTION
.I Mail
is intended as a replacement for the standard Bell mail program
(\fIbellmail(1)\fR), compatible with the Rand message-handling
system, mh.  When invoked without arguments, it simply 
execs 
.I inc(1) 
to incorporate new messages from the user's maildrop.
When one or more users is specified, a message is read from
the standard input and spooled to a temporary file.
.I Mail
then invokes
.I deliver(8) 
with the name of the temporary file as
its argument to deliver the message to the specified user.
.PP
The
.B subject
and
.B cc
switches can be used to insert the corresponding header components into
the message to be delivered.
.PP
This program is intended for the use of programs 
such as
.I at(1),
which expect
to automatically send mail to various users.
Normally, real people will prefer to use 
.I comp(1) 
and
.I send(1) 
to send messages.
.SH FILES
/etc/mh/deliver
.br
/tmp/mail*.tmp
.br
/usr/tmp/mh/lst*.tmp
.SH SEE ALSO
bellmail(1), rmail(1), send(1), inc(1), deliver(8)


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

docs/historical/mh-jun-1982/progs/mail.c
#ifdef COMMENT
	Proprietary Rand Corporation, 1981.
	Further distribution of this software
	subject to the terms of the Rand
	license agreement.
#endif

#include "../mh.h"
#include <stdio.h>
#include <signal.h>

extern char *sprintf();

#define NUMTOS  10              /* Number of to's & cc's accepted */

char    tmpfil[32];
int     exitstat = 1;
char    *subject, *body;

struct swit switches[] = {
	"body",         0,      /* 0 */
	"cc",           0,      /* 1 */
	"subject",      0,      /* 2 */
	"help",         4,      /* 3 */
	0,              0
};

main(argc, argv)
int argc;
char *argv[];
{
	register FILE *out;
	register int i, cnt;
	register char *cp;
	int pid, status, top, ccp, sig(), somebody = 0;
	char buf[BUFSIZ], *tos[NUMTOS], *ccs[NUMTOS], **argp;

	invo_name = argv[0];
	top = 0;
	ccp = -1;               /* -1 -> collecting TOs */
	if(argc == 1) {         /* Just call inc to read mail */
		execlp("inc", "inc", 0);
		perror("Mail: inc");
		done(exitstat);
	}
	argp = argv + 1;
	while(cp = *argp++) {
		if(*cp == '-')
			switch(smatch(++cp, switches)) {
			case -2:ambigsw(cp, switches);  /* ambiguous */
				done(exitstat);
							/* unknown */
			case -1:fprintf(stderr, "send: -%s unknown\n", cp);
				done(exitstat);
							/* -body */
			case 0: if((body = *argp++) == 0) {
		     missing:           fprintf(stderr, "Mail: Missing %s arg\n", cp);
					done(exitstat);
				}
				continue;
							/* -cc */
			case 1: ccp = 0;                /* Now collecting ccs */
				continue;
							/* -subject */
			case 2: if((subject = *argp++) == 0)
					goto missing;
				continue;
							/* -help */
			case 3: help("mail [switches] users ...",
				     switches);
				done(0);
			}
		else {
			if(ccp >= 0) {                  /* If getting ccs..*/
				if(ccp < NUMTOS)
					ccs[ccp++] = cp;
				else {
					fprintf(stderr, "Mail: Too many ccs\n");
					done(exitstat);
				}
			} else {                        /* Else, to */
				if(top < NUMTOS)
					tos[top++] = cp;
				else {
					fprintf(stderr, "Mail: Too many tos\n");
					done(exitstat);
				}
			}
		}
	}
					/* Create a mail temp file */
	VOID sprintf(tmpfil, "/tmp/%s", makename("mail", ".tmp"));
	if((out = fopen(tmpfil, "w")) == NULL) {
		perror(tmpfil);
		done(exitstat);
	}
	VOID signal(SIGINT, sig);       /* Clean up if user <del>s out */
	fprintf(out, "To: ");           /* Create to list */
	for(i = 0; i < top;) {
		fprintf(out, "%s", tos[i]);
		if(++i < top)
			fprintf(out, ", ");
	}
	fprintf(out, "\n");
	if(ccp > 0) {
		fprintf(out, "Cc: ");   /* Create cc list if needed */
		for(i = 0; i < ccp;) {
			fprintf(out, "%s", ccs[i]);
			if(++i < ccp)
				fprintf(out, ", ");
		}
		fprintf(out, "\n");
	}
	if(subject)                     /* Create subject if needed */
		fprintf(out, "Subject: %s\n", subject);

	fprintf(out, "\n");

	if(body) {                      /* Use body if I have it, */
		somebody++;
		fprintf(out, "%s\n", body);
	} else {                        /* Otherwise, get a body */
		while((cnt = read(0, buf, sizeof buf)) > 0) {
			somebody++;
			if(!fwrite(buf, cnt, 1, out)) {
				perror(tmpfil);
				sig();
			}
		}
	}

	if(ferror(out)) {               /* Check that all wrote well */
		fprintf(stderr, "Error writing tmp file\n");
		sig();
	}
	VOID fclose(out);
	if(!somebody)                   /* If NO body, then don't send */
		sig();                  /*  To be compatible with BELL mail */
	while((i = fork()) == -1) {     /* Now, deliver the mail */
		fprintf(stderr, "Waiting for a fork...\n");
		sleep(2);
	}
	if(i == 0) {                    /* Call deliverer in child */
		execl(mh_deliver, r1bindex(mh_deliver, '/'), tmpfil, 0);
		perror(mh_deliver);
		done(exitstat);
	}
	VOID signal(SIGINT, SIG_IGN);
	while((pid = wait(&status)) != -1 && pid != i) ;
	if(status) {                    /* And save mail if delivery failed */
		VOID signal(SIGINT, SIG_DFL);
		fprintf(stderr, "Letter saved in dead.letter\n");
		execl("/bin/mv", "mv", tmpfil, "dead.letter", 0);
		execl("/usr/bin/mv", "mv", tmpfil, "dead.letter", 0);
		perror("/bin/mv");
		done(exitstat);
	}
	exitstat = 0;
	sig();
}

sig()
{
	VOID unlink(tmpfil);
	done(exitstat);
}
